src/main/scala/org/neo4j/spark/service/SchemaService.scala

```diff
@@ -158,11 +168,178 @@ class SchemaService(private val options: Neo4jOptions, private val jobId: String
+        .map(_.quote())
+        .map(label => s"""
+             |MATCH (:$label)
+             |RETURN count(*) AS count""".stripMargin)
+        .mkString(" UNION ALL ")
+    } else {
+      queryReadStrategy.createStatementForNodeCount(options)
+    }
+    log.info(s"Executing the following counting query on Neo4j: $query")
+    session.run(query)
+      .list()
+      .asScala
+      .map(_.get("count"))
+      .map(count => if (count.isNull) 0L else count.asLong())
+      .min
+  }
+
+  def countForRelationshipWithQuery(filters: Array[Filter]): Long = {
+    val query = if (filters.isEmpty) {
+      val sourceQueries = options.relationshipMetadata.source.labels
+        .map(_.quote())
+        .map(label => s"""MATCH (:$label)-[${Neo4jUtil.RELATIONSHIP_ALIAS}:${options.relationshipMetadata.relationshipType}]->()
+                         |RETURN count(${Neo4jUtil.RELATIONSHIP_ALIAS}) AS count
+                         |""".stripMargin)
+      val targetQueries = options.relationshipMetadata.target.labels
+        .map(_.quote())
+        .map(label => s"""MATCH ()-[${Neo4jUtil.RELATIONSHIP_ALIAS}:${options.relationshipMetadata.relationshipType}]->(:$label)
+                         |RETURN count(${Neo4jUtil.RELATIONSHIP_ALIAS}) AS count
+                         |""".stripMargin)
+      (sourceQueries ++ targetQueries)
+        .mkString(" UNION ALL ")
+    } else {
+      queryReadStrategy.createStatementForRelationshipCount(options)
+    }
+    log.info(s"Executing the following counting query on Neo4j: $query")
+    session.run(query)
+      .list()
+      .asScala
+      .map(_.get("count"))
+      .map(count => if (count.isNull) 0L else count.asLong())
+      .min
+  }
+
+  def countForNode(filters: Array[Filter]): Long = try {
+    /*
+     * we try to leverage the count store in order to have the faster response possible
+     * https://neo4j.com/developer/kb/fast-counts-using-the-count-store/
+     * so in this scenario we have some limitations given the fact that we get the min
+     * for the sequence of counts returned
+     */
+    if (filters.isEmpty) {
+      val query = "CALL apoc.meta.stats() yield labels RETURN labels"
+      val map = session.run(query).single()
+        .asMap()
+        .asScala
+        .get("labels")
+        .getOrElse(Collections.emptyMap())
+        .asInstanceOf[util.Map[String, Long]].asScala
+      map.filterKeys(k => options.nodeMetadata.labels.contains(k))
+        .values.min
+    } else {
+      countForNodeWithQuery(filters)
+    }
+  } catch {
+    case e: ClientException =>
+      e.code match {
+        case "Neo.ClientError.Procedure.ProcedureNotFound" => {
+          countForNodeWithQuery(filters)
+        }
+        case _ => logExceptionForCount(e)
+      }
+    case e: Throwable => logExceptionForCount(e)
+  }
+
+  def countForRelationship(filters: Array[Filter]): Long = try {
+    if (filters.isEmpty) {
+      val query = "CALL apoc.meta.stats() yield relTypes RETURN relTypes"
```

Comment:
same as above db.relationshipTypes() ?

I thought we wanted to use apoc.meta.stats for the actual counts.